query/src/interpreters/async_insert_queue.rs

```diff
@@ -0,0 +1,419 @@
+use common_base::infallible::RwLock;
+use common_datablocks::DataBlock;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_planners::InsertPlan;
+use common_planners::SelectPlan;
+
+use super::InsertInterpreter;
+use super::SelectInterpreter;
+use crate::pipelines::new::executor::PipelineCompleteExecutor;
+use crate::pipelines::new::processors::port::InputPort;
+use crate::pipelines::new::processors::port::OutputPort;
+use crate::pipelines::new::processors::BlocksSource;
+use crate::pipelines::new::SinkPipeBuilder;
+use crate::pipelines::new::SourcePipeBuilder;
+use crate::sessions::QueryContext;
+use crate::sessions::SessionManager;
+use crate::sessions::SessionType;
+use crate::sessions::Settings;
+use crate::storages::memory::MemoryTableSink;
+
+#[derive(Clone)]
+pub struct InsertKey {
+    plan: Arc<InsertPlan>,
+    // settings different with default settings
+    changed_settings: Arc<Settings>,
+}
+
+impl InsertKey {
+    pub fn get_serialized_changed_settings(&self) -> String {
+        let mut serialized_settings = String::new();
+        let values = self.changed_settings.get_setting_values();
+        for value in values.into_iter() {
+            let serialized = serde_json::to_string(&value).unwrap();
+            serialized_settings.push_str(&serialized);
+        }
+        serialized_settings
+    }
+}
+
+impl PartialEq for InsertKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.plan.eq(&other.plan)
+            && self
+                .get_serialized_changed_settings()
+                .eq(&other.get_serialized_changed_settings())
+    }
+}
+
+impl Eq for InsertKey {}
+
+impl Hash for InsertKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let table = format!(
+            "{}.{}.{}",
+            self.plan.catalog_name, self.plan.database_name, self.plan.table_name
+        );
+        state.write(table.as_bytes());
+        let values = self.changed_settings.get_setting_values();
+        for value in values.into_iter() {
+            let serialized = serde_json::to_string(&value).unwrap();
+            state.write(serialized.as_bytes());
+        }
+    }
+}
+
+impl InsertKey {
+    pub fn try_create(plan: Arc<InsertPlan>, changed_settings: Arc<Settings>) -> Self {
+        Self {
+            plan,
+            changed_settings,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Entry {
+    block: DataBlock,
+    finished: Arc<RwLock<bool>>,
+    notify: Arc<Notify>,
+}
+
+impl Entry {
+    pub fn try_create(block: DataBlock) -> Self {
+        Self {
+            block,
+            finished: Arc::new(RwLock::new(false)),
+            notify: Arc::new(Notify::new()),
+        }
+    }
+
+    pub fn finish(&self) {
+        let mut finished = self.finished.write();
+        *finished = true;
+        self.notify.notify_one();
+    }
+
+    pub fn finish_with_timeout(&self) {
+        self.notify.notify_one();
+    }
+
+    pub async fn wait(&self) -> bool {
```

Comment:
Can we use https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html#method.wait_timeout ?

Reply:
We can use it. But I think `tokio::Notify` is more concise than condvar in use because it don't need to combined mutex and condvar.